### PR TITLE
18 refactoring

### PR DESCRIPTION
**Pull request by evgeniy-shpyra**

### CODE REVIEW

File: example.js [hunk 1]
```diff
@@ -24,105 +27,59 @@ const calculate = (packer, objects) => {
...
+
+    group.push({ w: item.w, h: item.h, d: item.d, count: item.count })
   }
-  const totalDimensions = calculateGroup(items)
 
-  console.timeEnd('start')
-  console.log('res: ', totalDimensions)
+  // calc parent group
+  const parentBin = calculate(packer, group)
+  if (!parentBin) throw new Error('Can\'t calculate a dimensions')
```
Comment (JaoodxD) [inline example.js]:
```suggestion
  if (!parentBin) throw new Error('Cannot calculate group dimensions')
```

File: experiments/compose.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
+const composeTheSame = (item, { maxCount }) => {
```
Comment (JaoodxD) [inline experiments/compose.js:1]:
```suggestion
const composeTheSame = (item, { maxCount } = {}) => {

File: example.js [hunk 3]
```diff
@@ -1,128 +1,81 @@
...
-  //   // items.push({ w: 21, h: 23, d:  12 }); // 100 / 5.6s
-  //   // items.push({ w: 21, h: 12, d:  23 }); // 100 / 3.4s
+    group.push({ w: item.w, h: item.h, d: item.d, count: item.count });
+  }
 
-//   // items.push({ w: 17, h: 26, d:  29 }); // 200 / 10.8s
-//   // items.push({ w: 17, h: 29, d:  26 }); // 200 / 0.8s
+  // calc parent group
+  const parentBin = calculate(packer, group);
+  if (!parentBin) throw new Error("Can't calculate a dimensions");
```
Comment (JaoodxD) [inline example.js:51]:
```suggestion
if (!parentBin) throw new Error('Cannot calculate group dimensions')
```

File: experiments/compose.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  const group = []
+  const subGroups = {}
+
+  const fullGroupsCount = Math.ceil(item.count / maxCount)
+  for (let i = 0; i < fullGroupsCount; i++) {
+    const isLastGroup = i === fullGroupsCount - 1
+    const groupCount =
+      isLastGroup && item.count % maxCount !== 0
+        ? item.count % maxCount
+        : maxCount
```
Comment (JaoodxD) [inline experiments/compose.js:11]: Complex ternary should be simplified or rewrited

File: experiments/compose.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
+    subGroups[groupKey].push(composedItem)
+  }
+
+  for (const key in subGroups) group.push(subGroups[key])
+
+  return group
+}
+
+const composeTheDifferent = (items, { maxCount }) => {
```
Comment (JaoodxD) [inline experiments/compose.js:27]:
```suggestion
const composeTheDifferent = (items, { maxCount } = {}) => {

File: experiments/compose.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  const groups = []
+  let subGroup = []
+  for (const item of items) {
+    if (subGroup.length >= maxCount) {
+      groups.push([...subGroup])
```
Comment (JaoodxD) [inline experiments/compose.js:32]:
```suggestion
      groups.push(subGroup)
```

Comment (JaoodxD) [reply experiments/compose.js:32]: Unnecessary array copy

File: experiments/compose.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  }
+  if (subGroup.length) {
+    groups.push(subGroup)
+  }
+  return groups
+}
+
+const getGroupVolume = (items) => {
+  let volume = 0
+  if (Array.isArray(items)) { for (const item of items) volume += getGroupVolume(item) }
```
Comment (JaoodxD) [inline experiments/compose.js:45]: Too complicated one liner

File: experiments/compose.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    subGroup.push(item)
+  }
+  if (subGroup.length) {
+    groups.push(subGroup)
+  }
+  return groups
+}
+
+const getGroupVolume = (items) => {
+  let volume = 0
```
Comment (JaoodxD) [inline experiments/compose.js:44]:
```suggestion
  let totalVolume = 0
```

File: experiments/compose.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  if (Array.isArray(items)) { for (const item of items) volume += getGroupVolume(item) }
+
+  for (const key in items) {
```
Comment (JaoodxD) [inline experiments/compose.js:47]:
```suggestion
  for (const key in items) {
    const volume = items[key].w * items[key].h * items[key].d
    const count = items[key].count || 1
    totalVolume += volume * count
  }
  ```

**Review by JaoodxD — CHANGES_REQUESTED** [review]

**Comment by evgeniy-shpyra** [comment]
fixed and add tests for packer.pack

File: test/compose.test.js [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+test('correct calculate groups volume, count = 1', async (t) => {
+  const inputData = [
+    { w: 10, h: 4, d: 1, count: 1 },
+    { w: 20, h: 10, d: 3, count: 1 },
+    { w: 12, h: 2, d: 3, count: 1 }
+  ]
+
+  const volume = getGroupVolume(inputData)
+
+  assert.equal(712, volume)
```
Comment (JaoodxD) [inline test/compose.test.js:14]: magic number 712. Not obvious where it comes from.

File: test/compose.test.js [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+test('correct calculate groups volume, count > 1 ', async (t) => {
+  const inputData = [
+    { w: 2, h: 4, d: 1, count: 10 },
+    { w: 20, h: 10, d: 3, count: 12 },
+    { w: 12, h: 2, d: 3, count: 2 }
+  ]
+
+  const volume = getGroupVolume(inputData)
+
+  assert.equal(7424, volume)
```
Comment (JaoodxD) [inline test/compose.test.js:26]: magic number

File: test/compose.test.js [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+      [
+        { w: 5, h: 3, d: 3, count: 1 },
+        { w: 4, h: 34, d: 3, count: 23 }
+      ]
+    ]
+  ]
+
+  const volume = getGroupVolume(inputData)
+
+  assert.equal(19823, volume)
```
Comment (JaoodxD) [inline test/compose.test.js:47]: magic number

File: test/compose.test.js [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+  for (let i = 0; i < itemsCount / maxTheSameCount; i++) {
+    expectedItems.push({
+      w: 10,
+      h: 4,
+      d: 1,
+      count: itemsCount / maxTheSameCount
+    })
+  }
+
+  assert.deepEqual([expectedItems], ...composedData)
```
Comment (JaoodxD) [inline test/compose.test.js:68]: Array of arrays `[expectedItems]` is intended?

File: test/main.test.js [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+  packer.createBins(200)
+  packer.sortBins()
+
+  const inputData = [
+    { w: 20, h: 10, d: 5, count: 1 },
+    { w: 5, h: 10, d: 20, count: 1 }
+  ]
+
+  const expectedDim = [20, 10, 10]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/main.test.js:17]: What a purpose to sort hard-coded array?

File: test/main.test.js [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+test('correct work calculate function: count = 2', async (t) => {
+  const packer = new Packer()
+  packer.createBins(200)
+  packer.sortBins()
+
+  const inputData = [{ w: 20, h: 10, d: 5, count: 2 }]
+
+  const expectedDim = [20, 10, 10]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/main.test.js:35]: Sort purpose?

File: test/main.test.js [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+  packer.createBins(200)
+  packer.sortBins()
+
+  const inputData = [
+    [{ w: 20, h: 10, d: 5, count: 1 }],
+    [{ w: 20, h: 10, d: 5, count: 1 }]
+  ]
+
+  const expectedDim = [20, 10, 10]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/main.test.js:56]: sort purpose?

File: test/main.test.js [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+test('correct work calculateGroup function: count > 1', async (t) => {
+  const packer = new Packer()
+  packer.createBins(200)
+  packer.sortBins()
+
+  const inputData = [[{ w: 20, h: 10, d: 5, count: 2 }]]
+
+  const expectedDim = [20, 10, 10]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/main.test.js:74]: sort purpose?

File: test/main.test.js [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+  packer.createBins(100)
+  packer.sortBins()
+
+  const inputData = [
+    [{ w: 10, h: 20, d: 5, count: 10 }],
+    { w: 10, h: 20, d: 5, count: 1 }
+  ]
+
+  const expectedDim = [25, 25, 20]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/main.test.js:95]: sort purpose?

File: test/main.test.js [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+  packer.createBins(100)
+  packer.sortBins()
+
+  const inputData = [
+    [{ w: 10, h: 20, d: 5, count: 10 }],
+    [{ w: 10, h: 20, d: 5, count: 10 }]
+  ]
+
+  const expectedDim = [40, 25, 20]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/main.test.js:116]: sort purpose?

File: test/packer.test.js [hunk 20]
```diff
@@ -31,3 +19,254 @@ test('cannot pack single item into smaller bin', async (t) => {
...
+
+test('correct pack single item', async (t) => {
+  const packer = new Packer()
+  packer.createBins(50)
+  packer.sortBins()
+
+  const inputItems = [{ w: 10, h: 11, d: 4 }]
+
+  const expectedDim = [10, 11, 4]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/packer.test.js:31]: sort purpose?

File: test/packer.test.js [hunk 21]
```diff
@@ -31,3 +19,254 @@ test('cannot pack single item into smaller bin', async (t) => {
...
+  packer.createBins(20)
+  packer.sortBins()
+
+  const inputItems = [
+    { w: 20, h: 10, d: 5 },
+    { w: 5, h: 10, d: 20 }
+  ]
+
+  const expectedDim = [20, 10, 10]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/packer.test.js:65]: sort purpose?

File: test/packer.test.js [hunk 22]
```diff
@@ -31,3 +19,254 @@ test('cannot pack single item into smaller bin', async (t) => {
...
+  packer.createBins(20)
+  packer.sortBins()
+
+  const inputItems = [
+    { w: 20, h: 20, d: 25 },
+    { w: 20, h: 20, d: 25 }
+  ]
+
+  const expectedDim = [20, 40, 25]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/packer.test.js:101]: sort purpose?

File: test/packer.test.js [hunk 23]
```diff
@@ -31,3 +19,254 @@ test('cannot pack single item into smaller bin', async (t) => {
...
+  packer.sortBins()
+
+  const inputItems = [
+    { w: 15, h: 10, d: 5 },
+    { w: 5, h: 10, d: 15 },
+    { w: 30, h: 5, d: 1 }
+  ]
+
+  const expectedDim = [30, 6, 10]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/packer.test.js:138]: sort purpose?

File: test/packer.test.js [hunk 24]
```diff
@@ -31,3 +19,254 @@ test('cannot pack single item into smaller bin', async (t) => {
...
+    { w: 12, h: 5, d: 1 },
+    { w: 12, h: 5, d: 1 },
+    { w: 30, h: 5, d: 1 },
+    { w: 10, h: 15, d: 5 },
+    { w: 10, h: 3, d: 5 },
+    { w: 5, h: 10, d: 12 }
+  ]
+
+  const expectedDim = [30, 6, 13]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/packer.test.js:178]: sort purpose?

File: test/packer.test.js [hunk 25]
```diff
@@ -31,3 +19,254 @@ test('cannot pack single item into smaller bin', async (t) => {
...
+    { w: 20, h: 10, d: 5 },
+    { w: 20, h: 10, d: 5 },
+    { w: 5, h: 10, d: 20 },
+    { w: 10, h: 20, d: 5 },
+    { w: 10, h: 5, d: 20 },
+    { w: 20, h: 10, d: 5 }
+  ]
+
+  const expectedDim = [25, 20, 20]
+  expectedDim.sort((a, b) => a - b)
```
Comment (JaoodxD) [inline test/packer.test.js:222]: sort purpose?

**Review by JaoodxD — CHANGES_REQUESTED** [review]

File: test/compose.test.js (hunk 13 above)
Comment (evgeniy-shpyra) [reply test/compose.test.js:68]: yes

**Review by JaoodxD — APPROVED** [review]
lgtm

**Pull request merged**